addons/product/models/product_template.py

```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
         value to input), indeed single value attributes can be used to filter
         products among others based on that attribute/value.
 
-        This method is necessary because it was previously possible to save a
-        line without any value on it, so the database might not be consistent in
-        that regard.
+        A product attribute value is considered valid for a template if it is
+        defined on a product template attribute line.
 
-        :return: all the valid product template attribute lines of this template
-        :rtype: recordset `product.template.attribute.line`
+        A product attribute is considered valid for a template if it
+        has at least one possible value set on the template.
+
+        For what is considered an archived variant, see `_has_valid_attributes`.
         """
-        self.ensure_one()
-        return self.attribute_line_ids.filtered(lambda ptal: ptal.value_ids)
+        archived_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', False)])
+
+        # prefetch
+        self.mapped('attribute_line_ids.value_ids.id')
+        self.mapped('attribute_line_ids.attribute_id.create_variant')
+
+        for record in self:
+            record.valid_product_template_attribute_line_ids = record.attribute_line_ids.filtered(lambda ptal: ptal.value_ids)
+            record.valid_product_template_attribute_line_wnva_ids = record.valid_product_template_attribute_line_ids._without_no_variant_attributes()
+
+            record.valid_product_attribute_value_ids = record.valid_product_template_attribute_line_ids.mapped('value_ids')
+            record.valid_product_attribute_value_wnva_ids = record.valid_product_template_attribute_line_wnva_ids.mapped('value_ids')
+
+            record.valid_product_attribute_ids = record.valid_product_template_attribute_line_ids.mapped('attribute_id')
+            record.valid_product_attribute_wnva_ids = record.valid_product_template_attribute_line_wnva_ids.mapped('attribute_id')
+
+            valid_value_ids = record.valid_product_attribute_value_wnva_ids
+            valid_attribute_ids = record.valid_product_attribute_wnva_ids
+
+            record.valid_archived_variant_ids = archived_variants.filtered(
+                lambda v: v.product_tmpl_id == record and v._has_valid_attributes(valid_attribute_ids, valid_value_ids)
+            )
 
     @api.multi
-    def _get_valid_product_attributes(self):
-        """A product attribute is considered valid for a template if it
-        has at least one possible value set on the template.
+    def _compute_valid_existing_variant_ids(self):
+        """This compute is done outside of `_compute_valid_attributes` because
+        it is often not needed, and it can be very bad on performance."""
+        existing_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', True)])
```

Comment:
active=true ? it is by default :/

Reply:
If someone put `active_test` in the context, I still want this method to make `active=True`, because that's what makes the field `valid_existing_variant_ids` different than `valid_archived_variant_ids`.